l3kernel/l3quark.dtx

```diff
@@ -574,6 +574,291 @@
+        \__kernel_msg_error:nnxx { kernel } { invalid-quark-function }
+          { \token_to_str:N #2 } {#1}
+        \use_none:nnn
+      }
+  }
+%    \end{macrocode}
+% \end{macro}
+% \end{macro}
+% \end{macro}
+%
+% \begin{macro}{
+%     \@@_test_generate_n:NNNn, \@@_test_generate_nn:NNNn,
+%     \@@_test_generate_N:NNNn, \@@_test_generate_Nn:NNNn,
+%     \@@_test_generate_NN:NNNn, \@@_test_generate_NN:NNNn,
+%   }
+%   These macros implement the six possibilities mentioned above, passing
+%   the right arguments to \cs{@@_test_generate_aux_do:nNNnnnn},
+%   which defines some auxiliaries, and then to
+%   \cs{@@_test_generate_define_tl:nNNNNN} (|:n(n)| variants) or to
+%   \cs{@@_test_generate_define_ifx:nNNNNN} (|:N(n)|) which define the
+%   main conditionals.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_test_generate_n:NNNn #1 #2 #3 #4
+  {
+    \@@_test_generate_aux_do:nNNnnnn {#4} #2 #3 { none } { } { } { }
+      \@@_test_generate_define_tl:nNNNNN #1 { }
+  }
+\cs_new_protected:Npn \@@_test_generate_nn:NNNn #1 #2 #3 #4
+  {
+    \@@_test_generate_aux_do:nNNnnnn {#4} #2 #3 { i } { n } {##1} {##2}
+      \@@_test_generate_define_tl:nNNNNN #1 { \use_none:n }
+  }
+\cs_new_protected:Npn \@@_test_generate_nN:NNNn #1 #2 #3 #4
+  {
+    \@@_test_generate_aux_do:nNNnnnn {#4} #2 #3 { i } { n } {##1} {##2}
+      \@@_test_generate_define_break_tl:nNNNNN #1 { }
+  }
+\cs_new_protected:Npn \@@_test_generate_N:NNNn #1 #2 #3 #4
+  {
+    \@@_test_generate_aux_do:nNNnnnn {#4} #2 #3 { none } { } { } { }
+      \@@_test_generate_define_ifx:nNNNNN #1 { }
+  }
+\cs_new_protected:Npn \@@_test_generate_Nn:NNNn #1 #2 #3 #4
+  {
+    \@@_test_generate_aux_do:nNNnnnn {#4} #2 #3 { i } { n } {##1} {##2}
+      \@@_test_generate_define_ifx:nNNNNN #1 { \use_none:n }
+  }
+\cs_new_protected:Npn \@@_test_generate_NN:NNNn #1 #2 #3 #4
+  {
+    \@@_test_generate_aux_do:nNNnnnn {#4} #2 #3 { i } { n } {##1} {##2}
+      \@@_test_generate_define_break_ifx:nNNNNN #1 { }
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{
+%     \@@_test_generate_aux_do:nNNnnnn,
+%     \@@_test_define_aux:NNNnnN
+%   }
+%   \cs{@@_test_generate_aux_do:nNNnnnn} makes the conrol sequence names
+%   which will be used by \cs{@@_test_define_aux:NNNnnN}, and then later
+%   by \cs{@@_test_generate_define_tl:nNNNNN} or
+%   \cs{@@_test_generate_define_ifx:nNNNNN}. The control sequences defined
+%   here are analogous to \cs{@@_if_recursion_tail:w} and to
+%   \cs[no-index]{use_(none|i)_delimit_by_q_recursion_stop:(|n)w}.
+%
+%   The name is composed by the name-space and the name of the quarks.
+%   Suppose \cs{__kernel_quark_test_generate:NNNn} was used with:
+%   \begin{verbatim}
+%     \__kernel_quark_test_generate:NNNn \__test_quark_tail:n
+%       \q__test_tail \q__test_stop { __test }
+%   \end{verbatim}
+%   then the first auxiliary will be \cs[no-index]{__test_if_test_tail:w},
+%   and the second one will be
+%   \cs[no-index]{__test_use_none_delimit_by_q_test_stop:w}.
```

Comment:
For the `recursion_tail` and `recursion_stop` perhaps (though it still
wouldn't allow dedicated quarks in a single module), but it would get
awkward for functions like `\quark_if_no_value:n(TF)` and
`\quark_if_nil:n(TF)` which are essentially the same, except for the
quark they test...